### PR TITLE
add @esri/hub-auth

### PR DESCRIPTION
**Pull request by jgravois**
```js
const session = checkSignInStatus()

if (session) // skip login!
```

to do:
- [x] trap for stale tokens/malformed sessions internally
- [ ] ?

### CODE REVIEW

File: packages/auth/src/auth.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  }
+  if (auth) {
+    const authJson = JSON.parse(decodeURIComponent(auth));
+    session = new UserSession({
+      token: authJson.token,
+      tokenExpires: new Date(authJson.expires)
+    });
+  }
+
+  // since we're isolating synchronous logic, is it even helpful to return a promise?
```
Comment (tomwayson) [inline packages/auth/src/auth.ts]: Yeah, I don't think we need a promise here.

File: packages/auth/src/auth.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  return new Promise((resolve, reject) => {
+    if (session) resolve(session);
+    else reject("No evidence of authentication found.");
```
Comment (tomwayson) [inline packages/auth/src/auth.ts]: If we don't return a promise, I'm not sure we should `throw` if no auth is found - maybe just return `null` and let the caller decide what to do from there.

File: packages/auth/test/auth.test.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  expires: TOMORROW,
+  allSSL: false,
+  accountId: "uCXeTVfooP4IIcx",
+  role: "account_admin",
+  urlKey: "subdomain",
+  customBaseUrl: "maps.arcgis.com"
+};
+
+/*
+  right now jasmine isn't running these tests in Node.js.
```
Comment (tomwayson) [inline packages/auth/test/auth.test.ts]: So, this comment made me think that `checkSignInStatus()` should if no `window` or `document`. 

Given that, it seems like what we want is to run the tests below only in a browser, and then add another test that runs only in node that verifies the above error. Is that possible w/ our test set up?

**Comment by jgravois** [comment]
```console
$ node --version
v8.11.1
$ npm --version
5.6.0
$ nvm --version
0.33.11
$ npm install 
npm ERR! code 1
npm ERR! Command failed: /usr/bin/git checkout a51eca7f8bafdc9bafab76d2305fedc64503304c
npm ERR! error: Your local changes to the following files would be overwritten by checkout:
npm ERR! 	.github/ISSUE_TEMPLATE/Bug_report.md
npm ERR! 	.github/ISSUE_TEMPLATE/Custom.md
npm ERR! 	.github/ISSUE_TEMPLATE/Feature_request.md
npm ERR! Please commit your changes or stash them before you switch branches.
npm ERR! Aborting
```
i've seen travis throw a lot of errors, but _complaining about local changes to non-existent files_ is a new one.

**Comment by tomwayson** [comment]
FYI - as per our discussion the other day, I think we should hold on this until we understand whether or not it is useful at all to check local storage for `torii-provider-arcgis`. A few reasons:
- that may only apply to apps that run on the .arcgis.com domain, so maybe the check should be done at the Ember addon level
- I want to understand https://github.com/Esri/torii-provider-arcgis/pull/43 and the subsequent PRs b/c @dbouwman preferred the cookie over the session auth, whereas this PR does the opposite

In short, we may not even need this package.

**Comment by jgravois** [comment]
to be continued...
